### ISSUE TITLE
Parameterized Filesystem Routing, e.g. `pages/product/@id.js`

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#207]
### Edit

We settled for this:

```
FILESYSTEM                 URL
pages/product/@id.js       /product/:id
```

### Original Post

Goal is to add support for Parameterized Filesytem Routing:

```
FILESYSTEM                 URL
pages/product/[pid].js     /product/:pid
```

On certain OS, e.g. Linux, filenames starting with a special charcater are poorly handled.

For example, on Linux, a file `[pid].js` needs the `[` character to be escaped, and `ls` doens't show the escape character but wraps the filename with `'`.
```
$ ls \[pid\].js 
'[pid].js'
```

Support for a filename like `:pid.js` is even worse on Linux.

What character could we use that is well supported across all OS?

**Comment by barelyhuman** [comment]
I'd go with `_` around the name, `_id_.js` , as long as we are limiting the language to be JS, if this is going to be used with Python , we might choose something else. 

or just `_id.js` works as well

**Comment by brillout** [comment]
> if this is going to be used with Python , we might choose something else.

How would that conflict? Isn't Python looking only for `_*.py` files? Wouldn't it ignore `_id.js`?

> `_id_.js`

We don't need advanced routes like `_id_some-suffix.js` since we can always define a Route String in `.page.route.js`.

But `_` is ugly, how about `!id.js` instead? I don't know about `!` support.

I guess `_id.js` is the best candidate so far.

**Comment by barelyhuman** [comment]
> How would that conflict? Isn't Python looking only for _*.py files? Wouldn't it ignore _id.js?

It would ignore the extension yes, my bad

> But _ is ugly, how about !id.js instead? I don't know about ! support.

I'll have to check as well

> I guess _id.js is the best candidate so far.

The other option can be `id.dyn.js` and that can be equally easy to parse but might not look nice

**Comment by brillout** [comment]
> The other option can be `id.dyn.js` and that can be equally easy to parse but might not look nice

Interesting idea

**Comment by brillout** [comment]
> I'll have to check as well

If `!` support is as good as `_`, then I'd go for `!`.

Or some other charachter. But can't think of an alternative to `!`.

**Comment by barelyhuman** [comment]
Write a small repo that handles all 3 extensions, add a github action to run on all 3 OS's , that just checks this. 
we'll have answers.

**Comment by brillout** [comment]
Sounds like too much effort. Let's stick to `_` for a while and see if someone else chimes in.

**Comment by barelyhuman** [comment]
I ran one on Linux and macOS, `!` failed, `_` seems to be working on both

**Comment by brillout** [comment]
(@barelyhuman Btw., if you didn't see that already, https://vercel.com/docs/file-system-api looks great. Can solve all the problems we were talking about.)

**Comment by brillout** [comment]
Alternatives to `_id.js`:
 - `$id.js`
 - `%id.js`
 - `@id.js`
 - `{id}.js`
 - `(id).js`
 - `` `id`.js``
 - `<id>.js`

**Comment by barelyhuman** [comment]
I think `{` works but creating it also requires the escape sequence `\[id\].js` | `\{id\}.js`, unless done from a code-editor, 
https://github.com/barelyhuman/ftrouter 
i do already have a similar router that uses square brackets though.

**Comment by brillout** [comment]
Yes `{id}.js` also needs escaping but, at least on my Linux terminal, the escaping is done automatically.

Also, upon `ls`, `{id}.js` is not wrapped with `'`. So it's better than `[id].js` in that regard.

Also `@id.js` works out perfectly on my terminal.

So my favorites so far are:
 - `@id.js`
 - `{id}.js`

Which are both a standard way of denoting a reference.

If `@id.js` is well supoprted cross-platform, I'd personally lean towards that.

**Comment by barelyhuman** [comment]
https://github.com/barelyhuman/file-checker-action-test/actions/runs/1471265605

Linux/Unix => everything other than `$` and `(` works 
Windows => everything other than `<` words. 

So i guess we can go ahead with `_`, `{}`, and `@`

**Comment by brillout** [comment]
Alright, I'm leaning towards `@id.js`. But I'm open for objections.

`{id}.js` is a close second.

I'll leave this open for some time and then implement.

This will introduce a breaking change; Route Strings should be `/product/@productId.js` instead of `/product/:productId.js`. 
 To stay coherent. A nice error message will help migration.

**Comment by brillout** [comment]
Debate about which one to choose going on Discord: https://discord.com/channels/815937377888632913/815937377888632916/910491401026928650

**Comment by brillout** [comment]
@barelyhuman How about `/pages/product/*.js` to also support catch-all routes? (Context: https://vite-plugin-ssr.com/catch-all.)

**Comment by brillout** [comment]
> @barelyhuman How about `/pages/product/*.js` to also support catch-all routes? (Context: https://vite-plugin-ssr.com/catch-all.)

Never mind, it doesn't work. (Like doing `vim *.js` opens all files ending with `.js`.)

**Comment by barelyhuman** [comment]
I think the catch all for next was using the same dyn approach, which would basically mean that if there's no other other params it would send everything to the same route. 


`/api/{x}.js` would catch all the routes that go `api/` 
but then if theres something like 

`/api/user.js` then that would go to that file when a request is made on `api/user` 

if I add a `user` folder and then add another dyn then that would handle sub routes for `api/user/123` 
`/api/user/{x}.js` would handle everything , other than `api/user`

So, you basically parse everything after that to the same context

**Comment by brillout** [comment]
So `api/some/deep/path` => `/api/{x}.js`?

**Comment by barelyhuman** [comment]
> So api/some/deep/path => /api/{x}.js?

Yes, and {x}.js get's the parameter as described in the docs but with a change that instead of it being `*` , it'll be `x` that has the values. 

So, you technically write a single matcher that matches in a downward fashion from the full string to the dynamic path 
and pass through accordingly. 

```js
pageContext.routeParams["x"] = {
 // whatever manipulation you do with `some/deep/path`
}
```

**Comment by brillout** [comment]
What's the difference to https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes then?

**Comment by barelyhuman** [comment]
I don't see a difference, they maintain 2 ways to define files which is `[param].js` thus creating a `params` object with that very key and in cases you are on a deeper dyn route like 

`api/[param1]/[param2]` then you have an `object` for `{param1:"value",param2:"value"}`  
and if you would do a `[....param].js` then it's just a single `{param:["value","value"]` for the same request

now why the differentiation, I don't know but the only advantage I see is it lets the developer know that he's written a catch-all route there and it's not just a dyn param route. 

Which makes sense but I don't think the matching logic for either would change much

**Comment by brillout** [comment]
Hm, AFAICT, it's confusing and I don't see the added value.

So far the plan:

`/product/42` => `/pages/product/@productId.js`
`/product/42/reviews` => 404

And with:

```js
// /pages/product/catch-all.page.route.js
export const '/product/*'
```

`/product/42/reviews` => `/pages/product/catch-all.page.route.js`

I like this most so far.

Catch-all routes are seldomly used and forcing the user to create a `.page.route.js` in order to implement a catch-all route is I'd say ok.

Objections, as always, welcome.

**Comment by brillout** [comment]
I will implement this after Telefunc's beta release.

**Comment by brillout** [comment]
Implemented in vite-plugin-ssr 0.4.
